File: leetcode/easy/valid-parentheses.py
# mine
# class Solution:
#     def isValid(self, s: str) -> bool:
#         stack = []
#         for c in s:
#             if c == '(':
#                 stack.append(')')
#             elif c == '{':
#                 stack.append('}')
#             elif c == '[':
#                 stack.append(']')
#             else:
#                 if len(stack) == 0 or c != stack.pop():
#                     return False
#         return len(stack) == 0

# ans
class Solution:
    def isValid(self, s: str) -> bool:
        d, stack = {
            '(': ')',
            '{': '}',
            '[': ']'
        }, []
        for c in s:
            if c in d:
                stack.append(c)
            elif len(stack) == 0 or d[stack.pop()] != c:
                return False
        return stack == []

File: leetcode/easy/valid-parentheses.ts
// https://leetcode.com/problems/valid-parentheses/

// MINE..
// function isValid(s: string): boolean {

// };

// ANSWER: stack + mirror
// var isValid = function (s: string): boolean {
//   const stack = [];

//   for (let i = 0; i < s.length; i++) {
//     let c = s.charAt(i);
//     switch (c) {
//       case '(':
//         stack.push(')');
//         break;
//       case '[':
//         stack.push(']');
//         break;
//       case '{':
//         stack.push('}');
//         break;
//       default:
//         if (c !== stack.pop()) {
//           return false;
//         }
//     }
//   }

//   return stack.length === 0;
// };

// mine..
// function isValid(s: string): boolean {
//     const map = {
//         '(': true,
//         '{': true,
//         '[': true,
//     }
//     for(let c of s) {
//         if(c === '(' || c === '{' || c === '[') map[c] = false
//         else if(map['('] && c === ')' || map['{'] && c === '}' || map['['] && c === ']') map[c] = false
//         else if(c === ')') map['('] = true
//         else if(c === '}') map['{'] = true
//         else map['['] = true
//         console.log({map})
//     }
//     for(let k in map) {
//         if(!map[k]) return false
//     }
//     return true
// };

// ans: stack
function isValid(s: string): boolean {
  const stack = []
  for (let c of s) {
    switch (c) {
      case '(':
        stack.push(')')
        break
      case '{':
        stack.push('}')
        break
      case '[':
        stack.push(']')
        break
      default:
        if (c !== stack.pop()) return false
    }
  }
  return !stack.length
}
